### DISCUSSION TITLE
Pitch Spellings, Intervals, and Key

### DISCUSSION THREAD

**Discussion by pikurasa** [sugarlabs/musicblocks-v4#22] [Ideas]
> And the code for generating the scales is much more robust. (https://github.com/sugarlabs/musicblocks-2/discussions/13#discussioncomment-275768)

I know I am literally a "day late and a dollar short" these days, but before it gets any later... here are my thoughts on scales. Please read and give feedback.

After a summer of working with Aviral on scales and modes, I question whether using this Whole-step, Half-Step method of determining scales is the right starting point.

I propose, instead, that we, for Alphabet spelling, focus on the Letter Class (LC) and Quality. By doing so, we can solve two problems. We can 1) maintain correct spellings, which may be important when using alternate temperaments. We can also 2) have the information we need to correctly measure intervals (which, themselves, have qualitative data such as Major, Minor, Augmented, and Diminished).

I realize that you have probably figured out how to correctly spell the notes (via Whole-step Half-step structure of scales) in all the cases you sent by email, but let's take an opportunity to consider a different approach and weigh the possible benefits.

Starting from Letter Class and Quality
======

I propose we start with Letter Class and Quality where Letter Class is A, B, C, D, E, F, or G. and quality is natural (), sharp (#), and flat (b). Natural is implied when no other quality is added (e.g. A is "A Natural"). "#" is Sharp and "b" is Flat.**

This, by itself, ensures that we have the correct spelling by default. Starting here, we can arrive at the correct Whole-step, Half-step structure, **making this the second step, not the first.**

Benefits
======

### Musical Meaning:
The spelling of a pitch usually indicates is meaning. Whereas F# in G Major indicates that we are on the leading tone of the scale, Gb in G Major would be a flag that something radical is going on as it would indicate we are moving down to F (Natural) which is outside of G Major (and thus we must be temporarily outside of G Major). Correct spellings for pitches give clues as to the musical context of a section of music.

This musical meaning will also be used to translate from LC to Scale Degree and/or Movable Solfege, whose spellings rely completely on musical context. (The mapping from LC and Quality to Fixed Do is 1:1. In other words La = A, Ti = B, Do = C, etc. so let's skip it for now as it is trivial.)

### Measuring Intervals:
In order to correctly measure the interval between two pitches, we need to know the quality of a pitch. For example, D to F# is a Major 3rd, whereas D to Gb is a Diminished 4th. First, the LC's are compared (i.e. D to F is a third), then the quality is determined (F Natural would be a Minor, whereas F# is a Major).

### Temperament:
By maintaining LC and Quality, we can move into different temperament accurately. For equal temperament F# and Gb will be equal, but for Just Intonation etc. F# will not be the same as Gb.

Other Angles Considered
========

### Should we start instead from Scale Degree or Moveable Do?
The other possibilities, which I have considered, are to start from a relative system instead, such as Scale Degree or Movable Do. I do not think we should start here as these rely on having a particular key strictly set from the beginning (as well as at all key changes) and music often has ambiguities (e.g. The beginning of Bethoven's Fifth when heard, not read) which make setting the key so strictly unwanted or impossible. Two prime examples are 1) temporary escapes from a key, and 2) "dovetail transitions". There is also music that is purposely atonal.

This all being said, having worked with Aviral on the many possibilities, I think **we should always have a seven pitch framework**, a kind of "Meta Scale" as our framework. I will write more about this later.

**Comment by walterbender** [comment]
C IONIAN [c d e f g a b c]
D DORIAN [d e f g a b c d]
E PHRYGIAN [e f g a b c d e]
F LYDIAN [f g a b c d e f]
G MIXOLYDIAN [g a b c d e f g]
A AEOLIAN [a b c d e f g a]
B LOCRIAN [b c d e f g a b]

G IONIAN [g a b c d e f# g]
A DORIAN [a b c d e f# g a]
B PHRYGIAN [b c d e f# g a b]
C LYDIAN [c d e f# g a b c]
D MIXOLYDIAN [d e f# g a b c d]
E AEOLIAN [e f# g a b c d e]
F# LOCRIAN [f# g a b c d e f#]

D IONIAN [d e f# g a b c# d]
E DORIAN [e f# g a b c# d e]
F# PHRYGIAN [f# g a b c# d e f#]
G LYDIAN [g a b c# d e f# g]
A MIXOLYDIAN [a b c# d e f# g a]
B AEOLIAN [b c# d e f# g a b]
C# LOCRIAN [c# d e f# g a b c#]

A IONIAN [a b c# d e f# g# a]
B DORIAN [b c# d e f# g# a b]
C# PHRYGIAN [c# d e f# g# a b c#]
D LYDIAN [d e f# g# a b c# d]
E MIXOLYDIAN [e f# g# a b c# d e]
F# AEOLIAN [f# g# a b c# d e f#]
G# LOCRIAN [g# a b c# d e f# g#]

E IONIAN [e f# g# a b c# d# e]
F# DORIAN [f# g# a b c# d# e f#]
G# PHRYGIAN [g# a b c# d# e f# g#]
A LYDIAN [a b c# d# e f# g# a]
B MIXOLYDIAN [b c# d# e f# g# a b]
C# AEOLIAN [c# d# e f# g# a b c#]
D# LOCRIAN [d# e f# g# a b c# d#]

B IONIAN [b c# d# e f# g# a# b]
C# DORIAN [c# d# e f# g# a# b c#]
D# PHRYGIAN [d# e f# g# a# b c# d#]
E LYDIAN [e f# g# a# b c# d# e]
F# MIXOLYDIAN [f# g# a# b c# d# e f#]
G# AEOLIAN [g# a# b c# d# e f# g#]
A# LOCRIAN [a# b c# d# e f# g# a#]

F# IONIAN [f# g# a# b c# d# e# f#]
G# DORIAN [g# a# b c# d# e# f# g#]
A# PHRYGIAN [a# b c# d# e# f# g# a#]
B LYDIAN [b c# d# e# f# g# a# b]
C# MIXOLYDIAN [c# d# e# f# g# a# b c#]
D# AEOLIAN [d# e# f# g# a# b c# d#]
E# LOCRIAN [e# f# g# a# b c# d# e#]

C# IONIAN [c# d# e# f# g# a# b# c#]
D# DORIAN [d# e# f# g# a# b# c# d#]
E# PHRYGIAN [e# f# g# a# b# c# d# e#]
F# LYDIAN [f# g# a# b# c# d# e# f#]
G# MIXOLYDIAN [g# a# b# c# d# e# f# g#]
A# AEOLIAN [a# b# c# d# e# f# g# a#]
B# LOCRIAN [b# c# d# e# f# g# a# b#]

F IONIAN [f g a bb c d e f]
G DORIAN [g a bb c d e f g]
A PHRYGIAN [a bb c d e f g a]
Bb LYDIAN [bb c d e f g a bb]
C MIXOLYDIAN [c d e f g a bb c]
D AEOLIAN [d e f g a bb c d]
E LOCRIAN [e f g a bb c d e]

Bb IONIAN [bb c d eb f g a bb]
C DORIAN [c d eb f g a bb c]
D PHRYGIAN [d eb f g a bb c d]
Eb LYDIAN [eb f g a bb c d eb]
F MIXOLYDIAN [f g a bb c d eb f]
G AEOLIAN [g a bb c d eb f g]
A LOCRIAN [a bb c d eb f g a]

Eb IONIAN [eb f g ab bb c d eb]
F DORIAN [f g ab bb c d eb f]
G PHRYGIAN [g ab bb c d eb f g]
Ab LYDIAN [ab bb c d eb f g ab]
Bb MIXOLYDIAN [bb c d eb f g ab bb]
C AEOLIAN [c d eb f g ab bb c]
D LOCRIAN [d eb f g ab bb c d]

Ab IONIAN [ab bb c db eb f g ab]
Bb DORIAN [bb c db eb f g ab bb]
C PHRYGIAN [c db eb f g ab bb c]
Db LYDIAN [db eb f g ab bb c db]
Eb MIXOLYDIAN [eb f g ab bb c db eb]
F AEOLIAN [f g ab bb c db eb f]
G LOCRIAN [g ab bb c db eb f g]

Db IONIAN [db eb f gb ab bb c db]
Eb DORIAN [eb f gb ab bb c db eb]
Fb PHRYGIAN [e f g a b c d e]
Gb LYDIAN [gb ab bb c db eb f gb]
Ab MIXOLYDIAN [ab bb c db eb f gb ab]
Bb AEOLIAN [bb c db eb f gb ab bb]
C LOCRIAN [c db eb f gb ab bb c]

Gb IONIAN [gb ab bb cb db eb f gb]
Ab DORIAN [ab bb cb db eb f gb ab]
Bb PHRYGIAN [bb cb db eb f gb ab bb]
Cb LYDIAN [cb db eb f gb ab bb cb]
Db MIXOLYDIAN [db eb f gb ab bb cb db]
Eb AEOLIAN [eb f gb ab bb cb db eb]
F LOCRIAN [f gb ab bb cb db eb f]

Cb IONIAN [cb db eb fb gb ab bb cb]
Db DORIAN [db eb fb gb ab bb cb db]
Eb PHRYGIAN [eb fb gb ab bb cb db eb]
Fb LYDIAN [fb gb ab bb cb db eb fb]
Gb MIXOLYDIAN [gb ab bb cb db eb fb gb]
Ab AEOLIAN [ab bb cb db eb fb gb ab]
Bb LOCRIAN [bb cb db eb fb gb ab bb]

I think these are all correct now.

**Reply by pikurasa** [reply]
These are correct.

The topic I tried to start here was referring to structure. If you have any questions about structure, let me know.

**Comment by walterbender** [comment]
C MAJOR PENTATONIC [c d e g a c]
G MAJOR PENTATONIC [g a b d e g]
D MAJOR PENTATONIC [d e f# a b d]
A MAJOR PENTATONIC [a b c# e f# a]
E MAJOR PENTATONIC [e f# g# b c# e]
B MAJOR PENTATONIC [b c# d# f# g# b]
F# MAJOR PENTATONIC [f# g# a# c# d# f#]
Db MAJOR PENTATONIC [db eb f ab bb db]
Ab MAJOR PENTATONIC [ab bb c eb f ab]
Eb MAJOR PENTATONIC [eb f g bb c eb]
Bb MAJOR PENTATONIC [bb c d f g bb]
F MAJOR PENTATONIC [f g a c d f]
C MINOR PENTATONIC [c eb f g bb c]
G MINOR PENTATONIC [g bb c d f g]
D MINOR PENTATONIC [d f g a c d]
A MINOR PENTATONIC [a c d e g a]
E MINOR PENTATONIC [e g a b d e]
B MINOR PENTATONIC [b d e f# a b]
F# MINOR PENTATONIC [f# a b c# e f#]
Db MINOR PENTATONIC [db e gb ab b db]
Ab MINOR PENTATONIC [ab b db eb gb ab]
Eb MINOR PENTATONIC [eb gb ab bb db eb]
Bb MINOR PENTATONIC [bb db eb f ab bb]
F MINOR PENTATONIC [f ab bb c eb f]
C WHOLE TONE [c d e f# g# a# c]
G WHOLE TONE [g a b c# d# f g]
D WHOLE TONE [d e f# g# a# c d]
A WHOLE TONE [a b c# d# f g a]
E WHOLE TONE [e f# g# a# c d e]
B WHOLE TONE [b c# d# f g a b]
F# WHOLE TONE [f# g# a# c d e f#]
Db WHOLE TONE [db eb f g a b db]
Ab WHOLE TONE [ab bb c d e gb ab]
Eb WHOLE TONE [eb f g a b db eb]
Bb WHOLE TONE [bb c d e gb ab bb]
F WHOLE TONE [f g a b db eb f]

Are these correct?

**Reply by pikurasa** [reply]
These are correct, but are these going to be tied to other scales?

What scales they are tied to will have implications for their scale degree assignments.

If you have questions let me know.

**Reply by walterbender** [reply]
The Major Pentatonic, Minor Pentatonic, and Whole Tone were generated by tying them to the corresponding scale in a 7-pitch framework. I can add other tie-ins as well as needed. (For example, I have also tied in Major and Minor Blues.

**Reply by pikurasa** [reply]
What corresponding 7-pitch framework did you use for Major Pentatonic, Minor Pentatonic, and Whole Tone?

At some point, we should start added the relative data points (e.g. Scale Degree, Movable Do).

**Comment by walterbender** [comment]
https://github.com/walterbender/musicblocks.js/blob/program-prototype/src/.prototype/keysignature.py#L279

**Comment by walterbender** [comment]
I made the following assumptions regarding the meantone temperaments:

(1) I assigned the following letter names to the 21 pitches:
```
ALL_NOTES = [
    "c",
    "c#",
    "db",
    "d",
    "d#",
    "eb",
    "e",
    "e#",
    "fb",
    "f",
    "f#",
    "gb",
    "g",
    "g#",
    "ab",
    "a",
    "a#",
    "bb",
    "b",
    "b#",
    "cb",
]
```

(2) I also tweaked the semi-tone transposition as follows:

C + 1 half step => C#
C + 2 half steps => D (not Db)

C - 1 half step => Cb (not B or B#)
C - 2 half steps => B (not B# or Bb)

Are these assumptions correct?

**Reply by pikurasa** [reply]
It should depend on what scale you are in.
